### PR TITLE
fix: issue 8511 The format of response body that returned by plugin should match Content-Type

### PR DESCRIPTION
**Pull request by MekelCon**
### Description
In the response.lua, improve management of the content type of the response.
If the response is a json, and if the original request accept it, we set the content-type to "application/json".
In other case we keep the old response type (text/plain)

Fixes [8511](https://github.com/apache/apisix/issues/8511)

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by spacewander** [comment]
Please make the CI pass, thanks!

**Comment by MekelCon** [comment]
> Please make the CI pass, thanks!

Should be done now !

File: apisix/core/response.lua [hunk 1]
```diff
@@ -80,7 +115,20 @@ function resp_exit(code, ...)
     end
 
     if idx > 0 then
-        ngx_print(concat_tab(t, "", 1, idx))
+        local response = concat_tab(t, "", 1, idx)
+        local _, err = decode_json(response)
+        if not err then
+            local content_type = ngx_header["Content-type"]
+            local accept_header = ngx.req.get_headers()["Accept"]
```
Comment (spacewander) [inline apisix/core/response.lua:122]: We can use https://github.com/apache/apisix/blob/c0c906eb4d62886ffa4b0938b017dc4810cfe639/apisix/core/request.lua#L104 which gets the header from cache

File: apisix/core/response.lua [hunk 2]
```diff
@@ -80,7 +115,20 @@ function resp_exit(code, ...)
     end
 
     if idx > 0 then
-        ngx_print(concat_tab(t, "", 1, idx))
+        local response = concat_tab(t, "", 1, idx)
+        local _, err = decode_json(response)
```
Comment (spacewander) [inline apisix/core/response.lua:119]: It would be more effective to check the header first before decoding, so we can save lots of unnecessary decoding.

Comment (MekelCon) [reply apisix/core/response.lua]: Fiexed by @MekelCon
[perf: decode json later](https://github.com/apache/apisix/pull/8587/commits/b4b60d28eb1ec0b1c0078633219d846e7be839a6)

File: apisix/core/response.lua (hunk 1 above)
Comment (MekelCon) [reply apisix/core/response.lua]: fixed by [get Accept header from cache](https://github.com/apache/apisix/pull/8587/commits/5e439faa537a97d825699059cc71e3b17df78aa9)

File: t/core/response.t [hunk 3]
```diff
@@ -199,3 +199,39 @@ aaa:
...
+GET /t
+--- error_code: 201
+--- response_body
+done
+--- response_headers
+Content-Type: text/plain
+
+
+
+=== TEST 10: exit with table then content-type is application/json
```
[the remark this answers is not in the document]
Comment (MekelCon) [reply t/core/response.t:223]: Fixed by [Add test : request not accepting application/json](https://github.com/apache/apisix/pull/8587/commits/e8d205a7b9d2467a72e38ee5b8e0b29b11bab23e)

**Comment by MekelCon** [comment]
@spacewander I saw the CI failed due to this line, 121 :
`local accept_header = core_request.header(ctx, "Accept")`
Because ctx is 'not defined'
Wich is true, but it works, i am a beginner in .lua so maybe there is something i am missing.

To have the correct syntax should we add the `ctx `parameter to the method resp_exit ?

**Comment by spacewander** [comment]
> @spacewander I saw the CI failed due to this line, 121 : `local accept_header = core_request.header(ctx, "Accept")` Because `ctx` is 'not defined' Wich is true, but it works, i am a beginner in .lua so maybe there is something i am missing.
> 
> To have the correct syntax should we add the `ctx `parameter to the method resp_exit ?

I think a bit it seems that there is no better way to do it. It's OK to add the parameter.

**Comment by MekelCon** [comment]
> > @spacewander I saw the CI failed due to this line, 121 : `local accept_header = core_request.header(ctx, "Accept")` Because `ctx` is 'not defined' Wich is true, but it works, i am a beginner in .lua so maybe there is something i am missing.
> > To have the correct syntax should we add the `ctx `parameter to the method resp_exit ?
> 
> I think a bit it seems that there is no better way to do it. It's OK to add the parameter.


@spacewander , 
so i checked quickly, 
i see at least 93 call to `response.exit(.......)`
SO before doing the update i want to be sure of what need to done.

As today we use this function with this definition : 
`function resp_exit(code, ...)` 

i must update it to :
`function resp_exit(ctx, code, ...)`

And then it means check all the usage to add the parameter ? For example
`return core.response.exit(502)`
will become
`return core.response.exit(ctx, 502)`

**Comment by spacewander** [comment]
> > > @spacewander I saw the CI failed due to this line, 121 : `local accept_header = core_request.header(ctx, "Accept")` Because `ctx` is 'not defined' Wich is true, but it works, i am a beginner in .lua so maybe there is something i am missing.
> > > To have the correct syntax should we add the `ctx `parameter to the method resp_exit ?
> > 
> > 
> > I think a bit it seems that there is no better way to do it. It's OK to add the parameter.
> 
> @spacewander , so i checked quickly, i see at least 93 call to `response.exit(.......)` SO before doing the update i want to be sure of what need to be done.
> 
> As today we use this function with this definition : `function resp_exit(code, ...)`
> 
> i must update it to : `function resp_exit(ctx, code, ...)`
> 
> And then it means check all the usage to add the parameter ? For example `return core.response.exit(502)` will become `return core.response.exit(ctx, 502)`

Yes, you are right

**Comment by MekelCon** [comment]
> > > > @spacewander I saw the CI failed due to this line, 121 : `local accept_header = core_request.header(ctx, "Accept")` Because `ctx` is 'not defined' Wich is true, but it works, i am a beginner in .lua so maybe there is something i am missing.
> > > > To have the correct syntax should we add the `ctx `parameter to the method resp_exit ?
> > > 
> > > 
> > > I think a bit it seems that there is no better way to do it. It's OK to add the parameter.
> > 
> > 
> > @spacewander , so i checked quickly, i see at least 93 call to `response.exit(.......)` SO before doing the update i want to be sure of what need to be done.
> > As today we use this function with this definition : `function resp_exit(code, ...)`
> > i must update it to : `function resp_exit(ctx, code, ...)`
> > And then it means check all the usage to add the parameter ? For example `return core.response.exit(502)` will become `return core.response.exit(ctx, 502)`
> 
> Yes, you are right

1st try done, i must say i am not very confident about what i did :/
At least i can confirm Apisix still run locally, and my test are passing.

File: apisix/init.lua [hunk 4]
```diff
@@ -372,7 +372,7 @@ function _M.handle_upstream(api_ctx, route, enable_websocket)
         local upstream = apisix_upstream.get_by_id(up_id)
         if not upstream then
             if is_http then
-                return core.response.exit(502)
+                return core.response.exit(core.ctx, 502)
```
Comment (spacewander) [inline apisix/init.lua:375]: Sorry, what does the core.ctx here mean? It should be `ctx` (or sometimes called `api_ctx`).

**Comment by soulbird** [comment]
Hi, @MekelCon Will you continue to complete this PR ?

**Comment by MekelCon** [comment]
> Hi, @MekelCon Will you continue to complete this PR ?

Yes i will !

**Pull request closed without merging**